File: heatmap/inst/htmlwidgets/heatmap.js
// NOTE http://www.htmlwidgets.org/develop_intro.html

HTMLWidgets.widget({

    name: 'heatmap',

    type: 'output',

    factory: function (el, width, height) {

        var chart = new Heatmap(el.id);

        // booleans converted from R may be true, false, 'TRUE', or 'FALSE'
        function getBool (bool) {
            return (bool === true || bool === 'TRUE');
        }

        return {
            renderValue: function (x) {
                var colClustOrder = x.options.colClustOrder;
                var rowClustOrder = x.options.rowClustOrder;
                var loColor = x.options.loColor;
                var hiColor = x.options.hiColor;
                var colorsBucket = x.options.colorsBucket;
                var dividersBucket = x.options.dividersBucket;
                var colCatScheme = x.options.colCatScheme;
                var colConScheme = x.options.colConScheme;
                var colAnnoHeatScheme = x.options.colAnnoHeatScheme;
                var rowCatScheme = x.options.rowCatScheme;
                var rowConScheme = x.options.rowConScheme;
                var rowAnnoHeatScheme = x.options.rowAnnoHeatScheme;
                var categorical = getBool(x.options.categorical);
                var scalingDim = x.options.scalingDim;
                var renderOnBrushEnd = getBool(x.options.renderOnBrushEnd);

                if (!chart.data) {
                    var data = HTMLWidgets.dataframeToD3(x.data);
                    var colnames = Object.keys(x.data).map(String);
                    var rowKey = colnames.shift();
                    var rownames = x.data[rowKey].map(String);

                    var matrix = rownames.map(function (rowname, j) {
                        return colnames.map(function (colname, k) {
                            return {
                                key: j + ' ' + k,
                                row: rowname,
                                col: colname,
                                value: x.data[colname][j]
                            };
                        });
                    });

                    var dataset = {
                        matrix: matrix,
                        rownames: rownames,
                        colnames: colnames
                    };

                    var options = {
                        colClustOrder: colClustOrder,
                        rowClustOrder: rowClustOrder,
                        loColor: loColor,
                        hiColor: hiColor,
                        colorsBucket: colorsBucket,
                        dividersBucket: dividersBucket,
                        colCatScheme: colCatScheme,
                        colConScheme: colConScheme,
                        colAnnoHeatScheme: colAnnoHeatScheme,
                        rowCatScheme: rowCatScheme,
                        rowConScheme: rowConScheme,
                        rowAnnoHeatScheme: rowAnnoHeatScheme,
                        categorical: categorical,
                        scalingDim: scalingDim,
                        renderOnBrushEnd: renderOnBrushEnd
                    };

                    chart.initialize(dataset, options);

                } else if (scalingDim !== chart.scalingDim) {

                    chart.updateColorScaling(scalingDim);

                } else {
                    var data = HTMLWidgets.dataframeToD3(x.data);
                    var colnames = Object.keys(x.data).map(String);
                    var rowKey = colnames.shift();
                    var rownames = x.data[rowKey].map(String);

                    var matrix = rownames.map(function (rowname, j) {
                        return colnames.map(function (colname, k) {
                            return {
                                key: j + ' ' + k,
                                row: rowname,
                                col: colname,
                                value: x.data[colname][j]
                            };
                        });
                    });

                    var dataset = {
                        matrix: matrix,
                        rownames: rownames,
                        colnames: colnames
                    };

                    var options = {
                        width: width,
                        height: height,
                        colClustOrder: colClustOrder,
                        rowClustOrder: rowClustOrder,
                        loColor: loColor,
                        hiColor: hiColor,
                        colorsBucket: colorsBucket,
                        dividersBucket: dividersBucket,
                        colCatScheme: colCatScheme,
                        colConScheme: colConScheme,
                        colAnnoHeatScheme: colAnnoHeatScheme,
                        rowCatScheme: rowCatScheme,
                        rowConScheme: rowConScheme,
                        rowAnnoHeatScheme: rowAnnoHeatScheme,
                        categorical: categorical,
                        scalingDim: scalingDim,
                        renderOnBrushEnd: renderOnBrushEnd
                    };

                    chart.initialize(dataset, options);
                }
            },

            resize: function (width, height) {
                chart.resize(width, height);
            }
        };
    }
});

File: heatmap/R/heatmap.R
#' @import htmlwidgets
#' @export

# You can learn more about package authoring with RStudio at:
#
#   http://r-pkgs.had.co.nz/
#
# Some useful keyboard shortcuts for package authoring:
#
#   Build and Reload Package:  'Cmd + Shift + B'
#   Check Package:             'Cmd + Shift + E'
#   Test Package:              'Cmd + Shift + T'
heatmap <- function (matrix,
                     width = NULL,
                     height = NULL,
                     colClustOrder = NULL,
                     rowClustOrder = NULL,
                     loColor = 'cornflowerblue',
                     hiColor = 'orange',
                     colorsBucket = c('red', 'orange', 'yellow', 'gray', 'cornflowerblue'),
                     dividersBucket = c(25, 50, 100, 500),
                     colCatScheme = 'rainbow',
                     colConScheme = 'rainbow',
                     colAnnoHeatScheme = 'plasma',
                     rowCatScheme = 'google',
                     rowConScheme = 'cubehelix',
                     rowAnnoHeatScheme = 'magma',
                     categorical = TRUE,
                     scalingDim = 'col',
                     renderOnBrushEnd = TRUE) {

    # read the matrix
    data <- matrix
    origColnames <- colnames(data)
    data <- cbind(rownames(data), data)
    colnames(data) <- c('X', origColnames)
    obj <- stats::heatmap(as.matrix(matrix))
    colClustOrder <- colnames(matrix)[obj$colInd]
    rowClustOrder <- rownames(matrix)[obj$rowInd][nrow(matrix):1]

    options <- list(
        id = 'heatmap',
        colClustOrder = colClustOrder,
        rowClustOrder = rowClustOrder,
        loColor = loColor,
        hiColor = hiColor,
        colorsBucket = colorsBucket,
        dividersBucket = dividersBucket,
        colCatScheme = colCatScheme,
        colConScheme = colConScheme,
        colAnnoHeatScheme = colAnnoHeatScheme,
        rowCatScheme = rowCatScheme,
        rowConScheme = rowConScheme,
        rowAnnoHeatScheme = rowAnnoHeatScheme,
        categorical = categorical,
        scalingDim = scalingDim,
        renderOnBrushEnd = renderOnBrushEnd
    )

    # pass the data and settings using 'x'
    x <- list(
        data = data,
        options = options
    )

    htmlwidgets::createWidget('heatmap', x, width = width, height = height)
}

#' @export
heatmapOutput <- function (outputId, width = '100%', height = '400px') {
    shinyWidgetOutput(outputId, 'heatmap', width, height, package = 'heatmap')
}

#' @export
renderHeatmap <- function (expr, env = parent.frame(), quoted = FALSE) {
    if (!quoted) { expr <- substitute(expr) } # force quoted
    shinyRenderWidget(expr, heatmapOutput, env, quoted = TRUE)
}
